core/blockchain.go

```diff
@@ -1480,13 +1462,18 @@ func (bc *BlockChain) insertChain(chain types.Blocks, verifySeals bool) (int, er
 		// Falls through to the block import
 	}
 	switch {
-	// First block is pruned, insert as sidechain and reorg only if TD grows enough
-	case errors.Is(err, consensus.ErrPrunedAncestor):
+	// We're in catalyst mode and the parent is pruned, try to recover the parent state
+	case catalystMode && errors.Is(err, consensus.ErrPrunedAncestor):
+		log.Debug("Pruned ancestor", "number", block.Number(), "hash", block.Hash())
+		return it.index, bc.recoverAncestors(block)
```

Comment:
How come we need to special case this? I find the flow here a bit hard to reason about. 
So somehow someone says "Import block 500". We're find that state for block 499 is missing, so we call `recoverAncestors. 
In recoverancestors, we build up hashes, 
```
	for parent != nil && !bc.HasState(parent.Root()) {
		hashes = append(hashes, parent.Hash())
		numbers = append(numbers, parent.NumberU64())
```
let's say hashes are  `[499, 498, 497, ... 420]`. 
Then, for each block, going from the end of the list, we do an `insertChain`. So `insertChain(420); insertChain(421)...`, ending on `insertChain(500)`. 

The `recoverAncestors` is a stripped-down version of `insertSideChain`. I'm just thinking, it would be neat if we could just reuse the existing methods we have. I'm guessing there's some good reason why we can't do that, in this case?